### ISSUE TITLE
YOLOE Visual Prompt based Classification

### ISSUE THREAD

**Issue by harshm2601** [ultralytics/ultralytics#24664]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Right now , it supports visual prompts for detection and that model can be exported. is it possible to visual prompt images and export as classifier? , As per my knowledge, it doesn't support classification natively but what would be the way if one needs to do that?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @harshm2601, thank you for your interest in Ultralytics 🚀! This is an automated response to help get your question routed quickly, and an Ultralytics engineer will also assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training or model capability ❓ Question, please provide as much detail as possible about your workflow, including:
- the exact model you are using
- whether you are working from Python, CLI, or exported formats
- what kind of visual prompt pipeline you are trying to build
- any logs, screenshots, or export steps attempted
- your `ultralytics` version and environment details

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by raimbekovm** [comment]
YOLOE doesn't ship a native classification task. `YOLOE.task_map` only registers `detect` and `segment`, so there's no classify head you can train or export directly off a YOLOE checkpoint. But the visual-prompt path gives you something equivalent for image-level labelling, and the classes survive export.

The idea is to define one bounding box per class on a reference image and pass it to `predict()` via `refer_image` + `visual_prompts`. YOLOE encodes each visual prompt through SAVPE into a class embedding and bakes it into the model via `set_classes` internally, so any subsequent predict on a new image detects (or segments) the prompted categories. Image-level "classification" of an input is then just argmax over the per-box confidence.

```python
import numpy as np
from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOEVPSegPredictor

model = YOLOE("yoloe-26s-seg.pt")
visual_prompts = dict(
    bboxes=np.array([[221.52, 405.8, 344.98, 857.54]]),
    cls=np.array([0]),
)

results = model.predict(
    "input.jpg",
    refer_image="reference.jpg",
    visual_prompts=visual_prompts,
    predictor=YOLOEVPSegPredictor,
)

boxes = results[0].boxes
image_class = int(boxes.cls[boxes.conf.argmax()]) if len(boxes) else None
```

A few gotchas: `cls` has to be integer IDs (np arrays of ints), strings will fail at runtime inside the visual-prompt transform. For seg checkpoints you need `predictor=YOLOEVPSegPredictor` explicitly, otherwise NMS hits a shape mismatch. And `model.names` ends up as generic `object0/object1/...` labels after the call, so override `model.names = {0: "your_label", ...}` if you want meaningful labels in the results.

Once `refer_image` is in play, the classes are persistent. Subsequent `predict()` calls don't need the prompts again, and the configured class set is baked into the exported weights. The full pattern is in the [YOLOE visual prompts section](https://docs.ultralytics.com/models/yoloe/#visual-prompt). One thing to keep in mind from the [export warning](https://docs.ultralytics.com/models/yoloe/#export-usage): once exported the model is static, so `set_classes()` and `visual_prompts=...` won't work on the export. To change classes you re-export from the `.pt` checkpoint.

**Comment by raimbekovm** [comment]
If what you actually want is a real classifier (single softmax head, no detection), YOLO and YOLOE don't share an architecture and there's no conversion path. The native option is [YOLO classify](https://docs.ultralytics.com/tasks/classify/):

```python
from ultralytics import YOLO

model = YOLO("yolo26n-cls.pt")
model.train(data="imagenet10", epochs=10)
model.export(format="onnx")
```

It uses a [folder-per-class dataset layout](https://docs.ultralytics.com/datasets/classify/) and exports cleanly to ONNX/TensorRT/TFLite via [the standard export modes](https://docs.ultralytics.com/modes/export/).

Rough rule of thumb: visual prompts when you have only a handful of example crops per class and detection-style output is fine. YOLO classify when you have many labelled examples per class and want a clean classifier export.

**Comment by rehan243** [comment]
Oh yeah, we hit something like this when we were trying to adapt YOLOv5 for basic classification tasks. You're right that YOLO doesn't natively support classification out of the box — it's just not what the architecture was designed for. But if you're sticking with YOLOE and want to force it into a classifier workflow, you can kinda hack around it by treating classification as a single-bounding-box detection problem.

Basically, you define a single box covering the entire image and set the class to whatever label you want. Then you can train it as usual. Here’s a quick example of what a dataset label might look like:

```yaml
annotations:
  - image: "image1.jpg"
    bbox: [0, 0, 640, 640]  # Full image
    class: "cat"
  - image: "image2.jpg"
    bbox: [0, 0, 640, 640]
    class: "dog"
```

Exports should work the same way since YOLO still thinks it's doing detection, but it'll just spit out one "detection" per image corresponding to your classification. Not perfect, but it works for quick hacks. 

Btw, if you’re planning to deploy this at scale, I'd maybe look at something purpose-built like EfficientNet. YOLO's overkill (and slower) for just classifying.

**Comment by glenn-jocher** [comment]
Small correction: that `annotations:` YAML isn’t a supported Ultralytics training format, and using one full-image box per image is still a detector workaround rather than YOLOE visual prompting or a true classifier. For prompted exports, use `YOLOE.predict(..., refer_image=..., visual_prompts=...)` as documented in the [`YOLOE.predict()` reference](https://docs.ultralytics.com/reference/models/yolo/model#ultralytics.models.yolo.model.YOLOE.predict); if you need an actual classification export, the supported path is a native `*-cls.pt` model as shown in the [classification predictor reference](https://docs.ultralytics.com/reference/models/yolo/classify/predict).

**Comment by rehan243** [comment]
yeah, fair point on the YAML not being an official format — that was more of a quick hack we used internally, definitely not a solid workflow. the `refer_image` param in `YOLOE.predict` is pretty slick for prompt-based stuff, hadn’t dug into that deeply before. good call on the native `*-cls.pt` models for real classification exports, that lines up with what we've seen too. you tried mixing visual prompts with classification exports yet? lmk if you wanna compare notes.

**Comment by glenn-jocher** [comment]
Not in the supported workflow: visual prompts are baked into `YOLOE` detect/segment exports, while native `*-cls` exports don’t accept `refer_image` or `visual_prompts`, as noted in the [YOLOE export docs](https://docs.ultralytics.com/models/yoloe#export-usage); if you have a concrete repro or blocker, please post the exact code and full traceback here and we’ll take a look.

**Comment by rehan243** [comment]
Yeah, that version issue definitely complicates things. I get that the visual prompts are tied to the exports, but if you're trying to mold `YOLOE` for classification, maybe check out the custom export scripts in the repo. There might be a way to bypass those restrictions, although it won’t be pretty. 

I’d be curious to see what code you’re working with, though. Any specific error messages popping up?

**Comment by glenn-jocher** [comment]
I wouldn’t recommend custom export-script bypasses here—there isn’t a supported path to turn `YOLOE` prompt exports into a native classifier, since the export remains a static detect/segment model with baked-in classes; if you need true image-level classification, the supported route is a `*-cls.pt` workflow from the [classification docs](https://docs.ultralytics.com/tasks/classify). If anyone hits a concrete failure, please share a minimal repro, the full traceback, and `yolo checks` output here.

**Comment by rehan243** [comment]
yeah, fair point on the export script — totally agree it’s not the cleanest path. hmm, I saw you’re on the latest YOLOE, right? If you’re aiming for a true classifier, those `*-cls.pt` workflows are definitely the way to go. curious if you’ve tried that route or hit any snags with it?

**Comment by raimbekovm** [comment]
@harshm2601

- **Visual prompts:** YOLOE has no native classify task, but `predict(refer_image=..., visual_prompts=...)` gives image-level labelling and the prompted classes are baked into the export. The image-level "class" is just argmax over the per-box confidence.
- **Real classifier:** if you need a true softmax classifier, use a native `*-cls.pt` model with a folder-per-class dataset, then export to ONNX/TensorRT/TFLite as usual.

Full code for both is in the comments above. Closing as resolved, feel free to reopen with a minimal repro if you hit a concrete blocker.
